File: web-client/twist.js
/**
 * Twist Protocol Module
 * =====================
 * 
 * Binary encoding/decoding for ROS2 geometry_msgs/Twist messages.
 * 
 * Binary Format (56 bytes, little-endian IEEE 754 double-precision):
 *   Offset  Size    Type      Field
 *   ─────────────────────────────────
 *   0       8       float64   linear.x
 *   8       8       float64   linear.y
 *   16      8       float64   linear.z
 *   24      8       float64   angular.x
 *   32      8       float64   angular.y
 *   40      8       float64   angular.z
 *   48      8       uint64    timestamp (ms since epoch)
 *   ─────────────────────────────────
 *   Total: 56 bytes
 * 
 * @module TwistProtocol
 */

// Message size constants
const TWIST_SIZE = 56;          // New size with timestamp
const TWIST_SIZE_LEGACY = 48;   // Legacy size without timestamp

/**
 * TwistMessage class representing ROS2 geometry_msgs/Twist
 */
class TwistMessage {
    /**
     * Create a TwistMessage
     * @param {Object} options - Velocity components
     * @param {number} [options.linearX=0] - Linear X velocity (m/s)
     * @param {number} [options.linearY=0] - Linear Y velocity (m/s)
     * @param {number} [options.linearZ=0] - Linear Z velocity (m/s)
     * @param {number} [options.angularX=0] - Angular X velocity (rad/s)
     * @param {number} [options.angularY=0] - Angular Y velocity (rad/s)
     * @param {number} [options.angularZ=0] - Angular Z velocity (rad/s)
     * @param {number} [options.timestamp=0] - Timestamp in ms (0 = auto-set on encode)
     */
    constructor({
        linearX = 0,
        linearY = 0,
        linearZ = 0,
        angularX = 0,
        angularY = 0,
        angularZ = 0,
        timestamp = 0
    } = {}) {
        this.linear = {
            x: linearX,
            y: linearY,
            z: linearZ
        };
        this.angular = {
            x: angularX,
            y: angularY,
            z: angularZ
        };
        this.timestamp = timestamp;
    }

    /**
     * Encode the Twist message to binary format
     * @param {boolean} [includeTimestamp=true] - Whether to include timestamp
     * @returns {ArrayBuffer} 56-byte binary representation
     */
    encode(includeTimestamp = true) {
        const size = includeTimestamp ? TWIST_SIZE : TWIST_SIZE_LEGACY;
        const buffer = new ArrayBuffer(size);
        const view = new DataView(buffer);
        
        // Little-endian encoding
        view.setFloat64(0, this.linear.x, true);
        view.setFloat64(8, this.linear.y, true);
        view.setFloat64(16, this.linear.z, true);
        view.setFloat64(24, this.angular.x, true);
        view.setFloat64(32, this.angular.y, true);
        view.setFloat64(40, this.angular.z, true);
        
        if (includeTimestamp) {
            // Set timestamp - if 0, use current time
            const ts = this.timestamp || Date.now();
            // Write as two 32-bit integers (JavaScript doesn't have native 64-bit int)
            view.setUint32(48, ts & 0xFFFFFFFF, true);        // Low 32 bits
            view.setUint32(52, Math.floor(ts / 0x100000000), true);  // High 32 bits
        }
        
        return buffer;
    }

    /**
     * Decode binary data into a TwistMessage
     * @param {ArrayBuffer} buffer - 48 or 56-byte binary data
     * @returns {TwistMessage} Decoded message
     * @throws {Error} If buffer size is invalid
     */
    static decode(buffer) {
        if (buffer.byteLength !== TWIST_SIZE && buffer.byteLength !== TWIST_SIZE_LEGACY) {
            throw new Error(
                `Invalid twist message size: expected ${TWIST_SIZE} or ${TWIST_SIZE_LEGACY} bytes, ` +
                `got ${buffer.byteLength} bytes`
            );
        }
        
        const view = new DataView(buffer);
        
        let timestamp = 0;
        if (buffer.byteLength >= TWIST_SIZE) {
            // Read timestamp as two 32-bit integers
            const low = view.getUint32(48, true);
            const high = view.getUint32(52, true);
            timestamp = low + (high * 0x100000000);
        }
        
        return new TwistMessage({
            linearX: view.getFloat64(0, true),
            linearY: view.getFloat64(8, true),
            linearZ: view.getFloat64(16, true),
            angularX: view.getFloat64(24, true),
            angularY: view.getFloat64(32, true),
            angularZ: view.getFloat64(40, true),
            timestamp: timestamp
        });
    }

    /**
     * Create a zero velocity message (stop/emergency stop command)
     * @returns {TwistMessage}
     */
    static zero() {
        return new TwistMessage({ timestamp: Date.now() });
    }

    /**
     * Create an emergency stop message (same as zero but explicit)
     * @returns {TwistMessage}
     */
    static emergencyStop() {
        return new TwistMessage({ timestamp: Date.now() });
    }

    /**
     * Create a forward motion command
     * @param {number} [speed=1.0] - Forward speed in m/s
     * @returns {TwistMessage}
     */
    static forward(speed = 1.0) {
        return new TwistMessage({ linearY: Math.abs(speed), timestamp: Date.now() });
    }

    /**
     * Create a backward motion command
     * @param {number} [speed=1.0] - Backward speed in m/s
     * @returns {TwistMessage}
     */
    static backward(speed = 1.0) {
        return new TwistMessage({ linearY: -Math.abs(speed), timestamp: Date.now() });
    }

    /**
     * Create a left turn command
     * @param {number} [rate=1.0] - Turn rate in rad/s
     * @returns {TwistMessage}
     */
    static turnLeft(rate = 1.0) {
        return new TwistMessage({ angularZ: Math.abs(rate), timestamp: Date.now() });
    }

    /**
     * Create a right turn command
     * @param {number} [rate=1.0] - Turn rate in rad/s
     * @returns {TwistMessage}
     */
    static turnRight(rate = 1.0) {
        return new TwistMessage({ angularZ: -Math.abs(rate), timestamp: Date.now() });
    }

    /**
     * Check if this is a stop command (all zeros)
     * @returns {boolean}
     */
    isZero() {
        return this.linear.x === 0 && this.linear.y === 0 && this.linear.z === 0 &&
               this.angular.x === 0 && this.angular.y === 0 && this.angular.z === 0;
    }

    /**
     * Calculate latency from this message's timestamp to now
     * @returns {number} Latency in milliseconds
     */
    getLatency() {
        if (this.timestamp === 0) return 0;
        return Date.now() - this.timestamp;
    }

    /**
     * Get string representation
     * @returns {string}
     */
    toString() {
        const lin = this.linear;
        const ang = this.angular;
        return `Twist{linear: [${lin.x.toFixed(3)}, ${lin.y.toFixed(3)}, ${lin.z.toFixed(3)}], ` +
               `angular: [${ang.x.toFixed(3)}, ${ang.y.toFixed(3)}, ${ang.z.toFixed(3)}], ` +
               `ts: ${this.timestamp}}`;
    }

    /**
     * Clone the message
     * @returns {TwistMessage}
     */
    clone() {
        return new TwistMessage({
            linearX: this.linear.x,
            linearY: this.linear.y,
            linearZ: this.linear.z,
            angularX: this.angular.x,
            angularY: this.angular.y,
            angularZ: this.angular.z,
            timestamp: this.timestamp
        });
    }
}

/**
 * Validate that data is a valid Twist message
 * @param {ArrayBuffer} buffer - Data to validate
 * @returns {boolean}
 */
function validateTwistData(buffer) {
    return buffer && (buffer.byteLength === TWIST_SIZE || buffer.byteLength === TWIST_SIZE_LEGACY);
}

/**
 * Quick encode function for direct values
 * @param {number} linearY - Linear Y velocity
 * @param {number} angularZ - Angular Z velocity
 * @returns {ArrayBuffer}
 */
function encodeTwist(linearY = 0, angularZ = 0) {
    return new TwistMessage({ linearY, angularZ, timestamp: Date.now() }).encode();
}

// Export for module usage
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        TwistMessage,
        TWIST_SIZE,
        TWIST_SIZE_LEGACY,
        validateTwistData,
        encodeTwist
    };
}

File: go-relay/twist.go
// Package twist provides binary encoding/decoding for ROS2 Twist messages.
//
// Binary Format (56 bytes, little-endian):
//   - Bytes 0-7:   linear.x  (float64)
//   - Bytes 8-15:  linear.y  (float64)
//   - Bytes 16-23: linear.z  (float64)
//   - Bytes 24-31: angular.x (float64)
//   - Bytes 32-39: angular.y (float64)
//   - Bytes 40-47: angular.z (float64)
//   - Bytes 48-55: timestamp (uint64, milliseconds since epoch)
package main

import (
	"encoding/binary"
	"errors"
	"fmt"
	"math"
	"time"
)

// TwistMessageSize is the fixed size of a binary Twist message in bytes.
const TwistMessageSize = 56

// TwistMessageSizeLegacy is the legacy size without timestamp.
const TwistMessageSizeLegacy = 48

// ErrInvalidMessageSize indicates the message size doesn't match expected format.
var ErrInvalidMessageSize = errors.New("invalid twist message size")

// Vector3 represents a 3D vector with X, Y, Z components.
// Used for both linear and angular velocity in Twist messages.
type Vector3 struct {
	X float64 `json:"x"` // X component
	Y float64 `json:"y"` // Y component
	Z float64 `json:"z"` // Z component
}

// TwistMessage represents a ROS2 geometry_msgs/Twist message.
// Contains linear and angular velocity vectors for robot motion control.
type TwistMessage struct {
	Linear    Vector3 `json:"linear"`    // Linear velocity (m/s)
	Angular   Vector3 `json:"angular"`   // Angular velocity (rad/s)
	Timestamp uint64  `json:"timestamp"` // Timestamp in milliseconds since epoch
}

// NewTwistMessage creates a new TwistMessage with zero velocities and current timestamp.
func NewTwistMessage() *TwistMessage {
	return &TwistMessage{
		Linear:    Vector3{X: 0, Y: 0, Z: 0},
		Angular:   Vector3{X: 0, Y: 0, Z: 0},
		Timestamp: uint64(time.Now().UnixMilli()),
	}
}

// EmergencyStop creates a zero-velocity message (emergency stop command).
func EmergencyStop() *TwistMessage {
	return NewTwistMessage()
}

// EncodeTwist converts a TwistMessage to its binary representation.
// Returns a 56-byte slice in little-endian format.
//
// Example:
//
//	twist := &TwistMessage{Linear: Vector3{Y: 1.0}, Angular: Vector3{Z: 0.5}}
//	data := EncodeTwist(twist)
func EncodeTwist(twist *TwistMessage) []byte {
	buf := make([]byte, TwistMessageSize)

	// Encode linear velocity vector (bytes 0-23)
	binary.LittleEndian.PutUint64(buf[0:8], math.Float64bits(twist.Linear.X))
	binary.LittleEndian.PutUint64(buf[8:16], math.Float64bits(twist.Linear.Y))
	binary.LittleEndian.PutUint64(buf[16:24], math.Float64bits(twist.Linear.Z))

	// Encode angular velocity vector (bytes 24-47)
	binary.LittleEndian.PutUint64(buf[24:32], math.Float64bits(twist.Angular.X))
	binary.LittleEndian.PutUint64(buf[32:40], math.Float64bits(twist.Angular.Y))
	binary.LittleEndian.PutUint64(buf[40:48], math.Float64bits(twist.Angular.Z))

	// Encode timestamp (bytes 48-55)
	// If timestamp is 0, use current time
	ts := twist.Timestamp
	if ts == 0 {
		ts = uint64(time.Now().UnixMilli())
	}
	binary.LittleEndian.PutUint64(buf[48:56], ts)

	return buf
}

// DecodeTwist parses a binary Twist message into a TwistMessage struct.
// Supports both 48-byte (legacy) and 56-byte (with timestamp) formats.
// Returns an error if the data size is invalid.
//
// Example:
//
//	twist, err := DecodeTwist(data)
//	if err != nil {
//	    log.Fatal(err)
//	}
//	fmt.Printf("Linear Y: %f\n", twist.Linear.Y)
func DecodeTwist(data []byte) (*TwistMessage, error) {
	if len(data) != TwistMessageSize && len(data) != TwistMessageSizeLegacy {
		return nil, fmt.Errorf("%w: expected %d or %d bytes, got %d bytes",
			ErrInvalidMessageSize, TwistMessageSize, TwistMessageSizeLegacy, len(data))
	}

	twist := &TwistMessage{}

	// Decode linear velocity vector (bytes 0-23)
	twist.Linear.X = math.Float64frombits(binary.LittleEndian.Uint64(data[0:8]))
	twist.Linear.Y = math.Float64frombits(binary.LittleEndian.Uint64(data[8:16]))
	twist.Linear.Z = math.Float64frombits(binary.LittleEndian.Uint64(data[16:24]))

	// Decode angular velocity vector (bytes 24-47)
	twist.Angular.X = math.Float64frombits(binary.LittleEndian.Uint64(data[24:32]))
	twist.Angular.Y = math.Float64frombits(binary.LittleEndian.Uint64(data[32:40]))
	twist.Angular.Z = math.Float64frombits(binary.LittleEndian.Uint64(data[40:48]))

	// Decode timestamp if present (bytes 48-55)
	if len(data) >= TwistMessageSize {
		twist.Timestamp = binary.LittleEndian.Uint64(data[48:56])
	}

	return twist, nil
}

// GetLatencyMs calculates the latency from the message timestamp to now.
// Returns latency in milliseconds.
func (t *TwistMessage) GetLatencyMs() int64 {
	if t.Timestamp == 0 {
		return 0
	}
	return time.Now().UnixMilli() - int64(t.Timestamp)
}

// String returns a human-readable representation of the TwistMessage.
func (t *TwistMessage) String() string {
	return fmt.Sprintf(
		"Twist{linear: [%.3f, %.3f, %.3f], angular: [%.3f, %.3f, %.3f], latency: %dms}",
		t.Linear.X, t.Linear.Y, t.Linear.Z,
		t.Angular.X, t.Angular.Y, t.Angular.Z,
		t.GetLatencyMs(),
	)
}

// IsZero returns true if all velocity components are zero.
func (t *TwistMessage) IsZero() bool {
	return t.Linear.X == 0 && t.Linear.Y == 0 && t.Linear.Z == 0 &&
		t.Angular.X == 0 && t.Angular.Y == 0 && t.Angular.Z == 0
}

// IsEmergencyStop is an alias for IsZero - checks if this is a stop command.
func (t *TwistMessage) IsEmergencyStop() bool {
	return t.IsZero()
}

// Clone creates a deep copy of the TwistMessage.
func (t *TwistMessage) Clone() *TwistMessage {
	return &TwistMessage{
		Linear:    Vector3{X: t.Linear.X, Y: t.Linear.Y, Z: t.Linear.Z},
		Angular:   Vector3{X: t.Angular.X, Y: t.Angular.Y, Z: t.Angular.Z},
		Timestamp: t.Timestamp,
	}
}
